File: apps/worker/src/public-api.ts
import { z } from "zod";
import {
  apiTokenCreateRequestSchema,
  apiTokenCreatePayloadSchema,
  apiTokenSchema,
  apiTokensListPayloadSchema,
  publicApiDueQuerySchema,
  publicApiDueItemSchema,
  publicApiDuePayloadSchema,
  publicApiMePayloadSchema,
  publicApiStatusPayloadSchema,
  publicApiSubscriptionPayloadSchema,
  publicApiSubscriptionsListPayloadSchema,
  publicApiSubscriptionsQuerySchema,
  publicApiTokenPlainSchema,
  type ApiToken,
} from "@renewlet/shared/schemas/public-api";
import { SUBSCRIPTION_STATUSES } from "@renewlet/shared/runtime";
import { requireAuth } from "./auth";
import { randomToken, sha256 } from "./crypto";
import {
  API_TOKEN_COLUMNS_FROM_API_TOKENS,
  SUBSCRIPTION_COLUMNS,
  countSubscriptions,
  getSettings,
  getSubscription,
  listApiTokenRows,
  listSubscriptionsPage,
  newId,
  nowIso,
  parseSubscriptionCursor,
  subscriptionCursor,
  toApiSubscription,
} from "./db";
import { dateOnlyInZone } from "./subscription-renewal";
import { bearerToken, HttpError, readJson, requestLocale, successJson } from "./http";
import { serverText } from "./server-i18n";
import type { ApiAppSettings, ApiTokenRow, Env, SubscriptionRow } from "./types";

const PUBLIC_API_TOKEN_PREFIX = "rlt_";
const PUBLIC_API_TOKEN_PREFIX_LENGTH = 12;
const PUBLIC_API_DUE_DEFAULT_DAYS = 30;

interface PublicApiAuth {
  userId: string;
  scopes: ["read"];
}

interface ApiTokenAuthRow extends ApiTokenRow {
  banned: number;
}

interface PublicApiSubscriptionsOptions {
  limit: number;
  cursor?: string;
  locale: ReturnType<typeof requestLocale>;
}

interface PublicApiDueOptions {
  settings?: ApiAppSettings;
}

export async function listApiTokens(request: Request, env: Env): Promise<Response> {
  const auth = await requireAuth(request, env);
  const tokens = (await listApiTokenRows(env, auth.user.id)).map(toApiToken);
  return noStoreSuccessJson(apiTokensListPayloadSchema.parse({ tokens }));
}

export async function createApiToken(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requireAuth(request, env);
  const body = await readJson(request, apiTokenCreateRequestSchema, locale);
  const timestamp = nowIso();
  const plainToken = publicApiTokenPlainSchema.parse(`${PUBLIC_API_TOKEN_PREFIX}${randomToken(32)}`);
  const tokenHash = await sha256(plainToken);
  const row: ApiTokenRow = {
    id: newId("tok"),
    user_id: auth.user.id,
    name: body.name,
    token_hash: tokenHash,
    token_prefix: plainToken.slice(0, PUBLIC_API_TOKEN_PREFIX_LENGTH),
    scopes_json: JSON.stringify(["read"]),
    last_used_at: null,
    created_at: timestamp,
    updated_at: timestamp,
  };
  // 明文 token 只返回本次响应；D1 写入前就收敛成 hash/prefix，后续列表无法复原明文。
  await env.DB.prepare(`
    INSERT INTO api_tokens (id, user_id, name, token_hash, token_prefix, scopes_json, last_used_at, created_at, updated_at)
    VALUES (?, ?, ?, ?, ?, ?, NULL, ?, ?)
  `).bind(row.id, row.user_id, row.name, row.token_hash, row.token_prefix, row.scopes_json, row.created_at, row.updated_at).run();
  return noStoreSuccessJson(apiTokenCreatePayloadSchema.parse({ token: toApiToken(row), plainToken }), { status: 201 });
}

export async function deleteApiToken(request: Request, env: Env, tokenId: string): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requireAuth(request, env);
  // 删除 token 是鉴权安全边界；D1 不保留墓碑行，旧 bearer 因 hash 行不存在立即 401。
  const result = await env.DB.prepare(`
    DELETE FROM api_tokens
    WHERE user_id = ? AND id = ?
  `).bind(auth.user.id, tokenId).run();
  if ((result.meta.changes ?? 0) === 0) throw new HttpError(404, serverText(locale, "common.notFound"), "NOT_FOUND");
  return noStoreSuccessJson({}, { status: 200 });
}

export async function publicApiMe(request: Request, env: Env): Promise<Response> {
  const auth = await requirePublicApiRead(request, env);
  return noStoreSuccessJson(publicApiMePayloadSchema.parse({ scopes: auth.scopes }));
}

export async function publicApiSubscriptions(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requirePublicApiRead(request, env);
  const url = new URL(request.url);
  const parsed = parseQuery(publicApiSubscriptionsQuerySchema, {
    limit: url.searchParams.get("limit") ?? undefined,
    cursor: url.searchParams.get("cursor") ?? undefined,
  }, locale);
  return noStoreSuccessJson(await readPublicApiSubscriptionsForUser(env, auth.userId, {
    limit: parsed.limit,
    ...(parsed.cursor ? { cursor: parsed.cursor } : {}),
    locale,
  }));
}

export async function publicApiSubscription(request: Request, env: Env, subscriptionId: string): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requirePublicApiRead(request, env);
  const row = await getSubscription(env, auth.userId, subscriptionId);
  if (!row) throw new HttpError(404, serverText(locale, "subscription.notFound"), "NOT_FOUND");
  return noStoreSuccessJson(publicApiSubscriptionPayloadSchema.parse({ subscription: toApiSubscription(row) }));
}

export async function publicApiStatus(request: Request, env: Env): Promise<Response> {
  const auth = await requirePublicApiRead(request, env);
  return noStoreSuccessJson(await readPublicApiStatusForUser(env, auth.userId));
}

export async function publicApiDue(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requirePublicApiRead(request, env);
  const url = new URL(request.url);
  const parsed = parseQuery(publicApiDueQuerySchema, {
    days: url.searchParams.get("days") ?? undefined,
  }, locale);
  return noStoreSuccessJson(await readPublicApiDueForUser(env, auth.userId, parsed.days ?? PUBLIC_API_DUE_DEFAULT_DAYS));
}

export async function readPublicApiSubscriptionsForUser(
  env: Env,
  userId: string,
  options: PublicApiSubscriptionsOptions,
): Promise<z.infer<typeof publicApiSubscriptionsListPayloadSchema>> {
  if (options.cursor && !parseSubscriptionCursor(options.cursor)) {
    throw new HttpError(400, serverText(options.locale, "common.invalidRequestParameters"), "INVALID_CURSOR");
  }
  const rows = await listSubscriptionsPage(env, userId, { limit: options.limit + 1, cursor: options.cursor });
  const pageRows = rows.slice(0, options.limit);
  const nextCursor = rows.length > options.limit ? subscriptionCursor(pageRows[pageRows.length - 1]!) : null;
  return publicApiSubscriptionsListPayloadSchema.parse({
    subscriptions: pageRows.map(toApiSubscription),
    nextCursor,
    total: await countSubscriptions(env, userId),
  });
}

export async function readPublicApiStatusForUser(env: Env, userId: string): Promise<z.infer<typeof publicApiStatusPayloadSchema>> {
  const byStatus = Object.fromEntries(SUBSCRIPTION_STATUSES.map((status) => [status, 0])) as Record<(typeof SUBSCRIPTION_STATUSES)[number], number>;
  const result = await env.DB.prepare(`
    SELECT status, COUNT(*) AS count
    FROM subscriptions
    WHERE user_id = ?
    GROUP BY status
  `).bind(userId).all<{ status: string; count: number }>();
  for (const row of result.results) {
    if (row.status in byStatus) byStatus[row.status as keyof typeof byStatus] = row.count;
  }
  const total = Object.values(byStatus).reduce((sum, count) => sum + count, 0);
  return publicApiStatusPayloadSchema.parse({
    generatedAt: nowIso(),
    total,
    byStatus,
  });
}

export async function readPublicApiDueForUser(env: Env, userId: string, days: number, options: PublicApiDueOptions = {}): Promise<z.infer<typeof publicApiDuePayloadSchema>> {
  const settings = options.settings ?? await getSettings(env, userId);
  const today = dateOnlyInZone(new Date(), settings.timezone);
  const through = addDateOnlyDays(today, days);
  const result = await env.DB.prepare(`
    SELECT ${SUBSCRIPTION_COLUMNS}
    FROM subscriptions
    WHERE user_id = ?
      AND (
        (next_billing_date >= ? AND next_billing_date <= ?)
        OR (trial_end_date >= ? AND trial_end_date <= ?)
      )
    ORDER BY next_billing_date ASC, trial_end_date ASC, created_at DESC, id DESC
  `).bind(userId, today, through, today, through).all<SubscriptionRow>();
  const items = result.results
    .map((row) => toDueItem(row, today, through))
    .filter((item): item is NonNullable<ReturnType<typeof toDueItem>> => item !== null)
    .sort((left, right) => left.dueDate.localeCompare(right.dueDate) || left.subscription.name.localeCompare(right.subscription.name));
  return publicApiDuePayloadSchema.parse({
    days,
    generatedAt: nowIso(),
    items,
  });
}

export async function readPublicApiNextDueForUser(env: Env, userId: string, options: PublicApiDueOptions = {}): Promise<z.infer<typeof publicApiDueItemSchema> | null> {
  const settings = options.settings ?? await getSettings(env, userId);
  const today = dateOnlyInZone(new Date(), settings.timezone);
  // Telegram /next 只需要第一条，但仍复用 Public API due item 契约；先按 owner 和未来日期缩小候选，再用同一 dueType 规则裁掉买断项。
  const result = await env.DB.prepare(`
    SELECT ${SUBSCRIPTION_COLUMNS}
    FROM subscriptions
    WHERE user_id = ?
      AND (
        (status = 'trial' AND trial_end_date IS NOT NULL AND trial_end_date >= ?)
        OR (next_billing_date >= ? AND (billing_cycle != 'one-time' OR one_time_term_count > 0))
      )
    ORDER BY
      CASE
        WHEN status = 'trial'
          AND trial_end_date IS NOT NULL
          AND trial_end_date >= ?
          AND (next_billing_date < ? OR trial_end_date <= next_billing_date)
        THEN trial_end_date
        ELSE next_billing_date
      END ASC,
      created_at DESC,
      id DESC
    LIMIT 10
  `).bind(userId, today, today, today, today).all<SubscriptionRow>();
  const items = result.results
    .map((row) => toDueItem(row, today, "9999-12-31"))
    .filter((item): item is z.infer<typeof publicApiDueItemSchema> => item !== null)
    .sort((left, right) => left.dueDate.localeCompare(right.dueDate) || left.subscription.name.localeCompare(right.subscription.name));
  return items[0] ?? null;
}

async function requirePublicApiRead(request: Request, env: Env): Promise<PublicApiAuth> {
  const locale = requestLocale(request);
  const token = bearerToken(request);
  const parsedToken = publicApiTokenPlainSchema.safeParse(token);
  if (!parsedToken.success) {
    throw new HttpError(401, serverText(locale, "auth.loginRequired"), "PUBLIC_API_UNAUTHORIZED");
  }
  const tokenHash = await sha256(parsedToken.data);
  const row = await env.DB.prepare(`
    SELECT ${API_TOKEN_COLUMNS_FROM_API_TOKENS}, users.banned AS banned
    FROM api_tokens
    JOIN users ON users.id = api_tokens.user_id
    WHERE api_tokens.token_hash = ?
    LIMIT 1
  `).bind(tokenHash).first<ApiTokenAuthRow>();
  if (!row || row.banned === 1 || !apiTokenHasReadScope(row)) {
    throw new HttpError(401, serverText(locale, "auth.loginRequired"), "PUBLIC_API_UNAUTHORIZED");
  }
  // Public API 与浏览器 session 分离；成功请求只刷新 token 使用时间，不延长或创建登录态。
  await env.DB.prepare("UPDATE api_tokens SET last_used_at = ?, updated_at = ? WHERE id = ?").bind(nowIso(), nowIso(), row.id).run();
  return { userId: row.user_id, scopes: ["read"] };
}

function toApiToken(row: ApiTokenRow): ApiToken {
  return apiTokenSchema.parse({
    id: row.id,
    name: row.name,
    tokenPrefix: row.token_prefix,
    scopes: ["read"],
    createdAt: row.created_at,
    lastUsedAt: row.last_used_at,
  });
}

function apiTokenHasReadScope(row: Pick<ApiTokenRow, "scopes_json">): boolean {
  try {
    const scopes = JSON.parse(row.scopes_json) as unknown;
    return Array.isArray(scopes) && scopes.includes("read");
  } catch {
    return false;
  }
}

function parseQuery<Schema extends z.ZodType>(schema: Schema, input: unknown, locale: ReturnType<typeof requestLocale>): z.infer<Schema> {
  const result = schema.safeParse(input);
  if (!result.success) {
    throw new HttpError(400, serverText(locale, "common.invalidRequestParameters"), "INVALID_QUERY", result.error.flatten());
  }
  return result.data;
}

function toDueItem(row: SubscriptionRow, today: string, through: string) {
  const dueType = dueTypeForSubscription(row, today, through);
  if (!dueType) return null;
  const subscription = toApiSubscription(row);
  return {
    dueDate: dueType === "trial" ? row.trial_end_date! : row.next_billing_date,
    dueType,
    subscription,
  };
}

function dueTypeForSubscription(row: SubscriptionRow, today: string, through: string): "renewal" | "trial" | "expiry" | null {
  if (row.status === "trial" && row.trial_end_date && row.trial_end_date >= today && row.trial_end_date <= through) return "trial";
  if (row.next_billing_date < today || row.next_billing_date > through) return null;
  if (row.billing_cycle === "one-time") return row.one_time_term_count && row.one_time_term_count > 0 ? "expiry" : null;
  return "renewal";
}

function addDateOnlyDays(value: string, days: number): string {
  const [yearText, monthText, dayText] = value.split("-");
  const year = Number.parseInt(yearText ?? "", 10);
  const month = Number.parseInt(monthText ?? "", 10);
  const day = Number.parseInt(dayText ?? "", 10);
  if (!Number.isInteger(year) || !Number.isInteger(month) || !Number.isInteger(day)) return value;
  return new Date(Date.UTC(year, month - 1, day + days)).toISOString().slice(0, 10);
}

function noStoreSuccessJson(value: unknown, init: ResponseInit = {}): Response {
  const response = successJson(value, init);
  response.headers.set("cache-control", "no-store");
  response.headers.set("x-content-type-options", "nosniff");
  return response;
}

File: apps/docker-server/cmd/renewlet/public_api.go
package main

// public_api.go 实现通用只读 Public API。
//
// 架构位置：
//   - 登录态 `/api/app/api-tokens` 只负责 token 生命周期管理。
//   - 公开 `/api/public/v1/*` 只接受 `Authorization: Bearer rlt_*`，不读取 session、日历 token 或公开页 token。
//   - 订阅响应复用站内订阅 DTO，避免 Telegram/CLI/Shortcuts 后续各自读库。
import (
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

const (
	publicAPITokenRandomBytes  = 32
	publicAPITokenPrefix       = "rlt_"
	publicAPITokenPrefixLength = 12
	publicAPIDueDefaultDays    = 30
	publicAPIDueMaxDays        = 366
	publicAPIDueResultLimit    = 500
)

var publicAPITokenRe = regexp.MustCompile(`^rlt_[A-Za-z0-9_-]{43}$`)
var bearerTokenHeaderRe = regexp.MustCompile(`(?i)^Bearer\s+(.+)$`)
var errPublicAPIInvalidCursor = errors.New("invalid public api cursor")

type apiTokenDTO struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	TokenPrefix string   `json:"tokenPrefix"`
	Scopes      []string `json:"scopes"`
	CreatedAt   string   `json:"createdAt"`
	LastUsedAt  *string  `json:"lastUsedAt,omitempty"`
}

type apiTokensListResponse struct {
	Tokens []apiTokenDTO `json:"tokens"`
}

type apiTokenCreateRequest struct {
	Name string `json:"name"`
}

func (r *apiTokenCreateRequest) Validate(locale appLocale) error {
	r.Name = strings.TrimSpace(r.Name)
	if r.Name == "" || len([]rune(r.Name)) > 80 {
		return errors.New(serverText(locale, "common.invalidRequestParameters"))
	}
	return nil
}

type apiTokenCreateResponse struct {
	Token      apiTokenDTO `json:"token"`
	PlainToken string      `json:"plainToken"`
}

type publicAPIAuthContext struct {
	UserID string
	Scopes []string
}

type publicAPIMeResponse struct {
	Scopes []string `json:"scopes"`
}

type publicAPIStatusResponse struct {
	GeneratedAt string           `json:"generatedAt"`
	Total       int64            `json:"total"`
	ByStatus    map[string]int64 `json:"byStatus"`
}

type publicAPIDueItem struct {
	DueDate      string                 `json:"dueDate"`
	DueType      string                 `json:"dueType"`
	Subscription map[string]interface{} `json:"subscription"`
}

type publicAPIDueResponse struct {
	Days        int                `json:"days"`
	GeneratedAt string             `json:"generatedAt"`
	Items       []publicAPIDueItem `json:"items"`
}

func handleAPITokensList(app core.App, e *core.RequestEvent) error {
	rows, err := app.FindRecordsByFilter("api_tokens", "user = {:user}", "-created,-id", 200, 0, dbx.Params{"user": e.Auth.Id})
	if err != nil {
		return e.InternalServerError(serverText(requestLocale(e.Request), "common.internalError"), err)
	}
	tokens := make([]apiTokenDTO, 0, len(rows))
	for _, row := range rows {
		tokens = append(tokens, apiTokenDTOFromRecord(row))
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, apiTokensListResponse{Tokens: tokens})
}

func handleAPITokenCreate(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	body, err := decodeStrictJSON[apiTokenCreateRequest](e.Request, locale)
	if err != nil {
		return e.BadRequestError(validationErrorMessage(locale, "common.invalidRequestBody", err), err)
	}
	record, plainToken, err := createAPITokenRecord(app, e.Auth.Id, body.Name)
	if err != nil {
		return e.InternalServerError(serverText(locale, "common.internalError"), err)
	}
	// 明文 token 只在创建响应出现一次；列表和数据库都只保留 prefix/hash。
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusCreated, apiTokenCreateResponse{
		Token:      apiTokenDTOFromRecord(record),
		PlainToken: plainToken,
	})
}

func handleAPITokenDelete(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	record, err := findAPITokenForUser(app, e.Auth.Id, strings.TrimSpace(e.Request.PathValue("id")))
	if err != nil {
		return e.NotFoundError(serverText(locale, "common.notFound"), err)
	}
	// 删除 token 是 Public API 鉴权安全边界；hash 行消失后，外部保存的旧明文只能得到 401。
	if err := app.Delete(record); err != nil {
		return e.InternalServerError(serverText(locale, "common.internalError"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiEmptySuccessJSON(e, http.StatusOK)
}

func handlePublicAPIMe(app core.App, e *core.RequestEvent) error {
	auth, err := authenticatePublicAPIRequest(app, e)
	if err != nil {
		return publicAPIAuthError(e, err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, publicAPIMeResponse{Scopes: auth.Scopes})
}

func handlePublicAPISubscriptionsList(app core.App, e *core.RequestEvent) error {
	auth, err := authenticatePublicAPIRequest(app, e)
	if err != nil {
		return publicAPIAuthError(e, err)
	}
	locale := requestLocale(e.Request)
	limit, err := parsePositiveQueryInt(e.Request.URL.Query().Get("limit"), 50, 1, 100)
	if err != nil {
		return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), err)
	}
	response, err := publicAPISubscriptionsForUser(app, auth.UserID, limit, e.Request.URL.Query().Get("cursor"))
	if err != nil {
		if errors.Is(err, errPublicAPIInvalidCursor) {
			return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), err)
		}
		return e.InternalServerError(serverText(locale, "common.internalError"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, response)
}

func handlePublicAPISubscriptionDetail(app core.App, e *core.RequestEvent) error {
	auth, err := authenticatePublicAPIRequest(app, e)
	if err != nil {
		return publicAPIAuthError(e, err)
	}
	record, err := findSubscriptionForPublicAPI(app, auth.UserID, strings.TrimSpace(e.Request.PathValue("id")))
	if err != nil {
		return e.NotFoundError(serverText(requestLocale(e.Request), "subscription.notFound"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, subscriptionResponse{Subscription: subscriptionAPIFromRecord(record)})
}

func handlePublicAPIStatus(app core.App, e *core.RequestEvent) error {
	auth, err := authenticatePublicAPIRequest(app, e)
	if err != nil {
		return publicAPIAuthError(e, err)
	}
	response, err := publicAPIStatusForUser(app, auth.UserID)
	if err != nil {
		return e.InternalServerError(serverText(requestLocale(e.Request), "common.internalError"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, response)
}

func handlePublicAPIDue(app core.App, e *core.RequestEvent) error {
	auth, err := authenticatePublicAPIRequest(app, e)
	if err != nil {
		return publicAPIAuthError(e, err)
	}
	locale := requestLocale(e.Request)
	days, err := parsePositiveQueryInt(e.Request.URL.Query().Get("days"), publicAPIDueDefaultDays, 1, publicAPIDueMaxDays)
	if err != nil {
		return e.BadRequestError(serverText(locale, "common.invalidRequestParameters"), err)
	}
	response, err := publicAPIDueForUser(app, auth.UserID, days)
	if err != nil {
		return e.InternalServerError(serverText(locale, "common.internalError"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiSuccessJSON(e, http.StatusOK, response)
}

func publicAPISubscriptionsForUser(app core.App, userID string, limit int, rawCursor string) (subscriptionsListResponse, error) {
	filter := "user = {:user}"
	params := dbx.Params{"user": userID}
	if cursorText := strings.TrimSpace(rawCursor); cursorText != "" {
		cursor, err := parseSubscriptionCursorPayload(cursorText)
		if err != nil {
			return subscriptionsListResponse{}, errPublicAPIInvalidCursor
		}
		filter = "user = {:user} && (created < {:createdAt} || (created = {:createdAt} && id < {:id}))"
		params["createdAt"] = cursor.CreatedAt
		params["id"] = cursor.ID
	}
	rows, err := app.FindRecordsByFilter("subscriptions", filter, "-created,-id", limit+1, 0, params)
	if err != nil {
		return subscriptionsListResponse{}, err
	}
	pageRows := rows
	var nextCursor *string
	if len(rows) > limit {
		pageRows = rows[:limit]
		cursor := encodeSubscriptionCursor(pageRows[len(pageRows)-1])
		nextCursor = &cursor
	}
	subscriptions := make([]map[string]interface{}, 0, len(pageRows))
	for _, record := range pageRows {
		subscriptions = append(subscriptions, subscriptionAPIFromRecord(record))
	}
	total, err := app.CountRecords("subscriptions", dbx.HashExp{"user": userID})
	if err != nil {
		return subscriptionsListResponse{}, err
	}
	return subscriptionsListResponse{Subscriptions: subscriptions, NextCursor: nextCursor, Total: total}, nil
}

func publicAPIStatusForUser(app core.App, userID string) (publicAPIStatusResponse, error) {
	byStatus := map[string]int64{
		"trial":     0,
		"active":    0,
		"expired":   0,
		"paused":    0,
		"cancelled": 0,
	}
	var total int64
	for status := range byStatus {
		count, err := app.CountRecords("subscriptions", dbx.HashExp{"user": userID, "status": status})
		if err != nil {
			return publicAPIStatusResponse{}, err
		}
		byStatus[status] = count
		total += count
	}
	return publicAPIStatusResponse{
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
		Total:       total,
		ByStatus:    byStatus,
	}, nil
}

func publicAPIDueForUser(app core.App, userID string, days int) (publicAPIDueResponse, error) {
	settings := publicStatusSettingsForUser(app, userID)
	return publicAPIDueForUserWithSettings(app, userID, days, settings)
}

func publicAPIDueForUserWithSettings(app core.App, userID string, days int, settings appSettings) (publicAPIDueResponse, error) {
	today := todayDateOnly(time.Now().UTC(), settings.Timezone)
	through := addDateOnly(today, days)
	rows, err := app.FindRecordsByFilter(
		"subscriptions",
		"user = {:user} && ((nextBillingDate >= {:today} && nextBillingDate <= {:through}) || (trialEndDate >= {:today} && trialEndDate <= {:through}))",
		"nextBillingDate,trialEndDate,-created,-id",
		// Public API due 是摘要入口而非导出接口；结果设硬上限，避免外部轮询把单次响应放大。
		publicAPIDueResultLimit,
		0,
		dbx.Params{"user": userID, "today": today, "through": through},
	)
	if err != nil {
		return publicAPIDueResponse{}, err
	}
	return publicAPIDueResponse{
		Days:        days,
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
		Items:       publicAPIDueItemsFromRecords(rows, today, through),
	}, nil
}

func publicAPINextDueForUserWithSettings(app core.App, userID string, settings appSettings) (*publicAPIDueItem, error) {
	today := todayDateOnly(time.Now().UTC(), settings.Timezone)
	// Telegram /next 只需要第一条，但仍复用 Public API due item 契约；先按 owner 和未来日期缩小候选，再用同一 dueType 规则裁掉买断项。
	rows, err := app.FindRecordsByFilter(
		"subscriptions",
		"user = {:user} && ((status = 'trial' && trialEndDate >= {:today}) || (nextBillingDate >= {:today} && (billingCycle != 'one-time' || oneTimeTermCount > 0)))",
		"nextBillingDate,trialEndDate,-created,-id",
		publicAPIDueResultLimit,
		0,
		dbx.Params{"user": userID, "today": today},
	)
	if err != nil {
		return nil, err
	}
	items := publicAPIDueItemsFromRecords(rows, today, "9999-12-31")
	sort.SliceStable(items, func(i, j int) bool {
		if items[i].DueDate == items[j].DueDate {
			return publicAPISubscriptionName(items[i].Subscription) < publicAPISubscriptionName(items[j].Subscription)
		}
		return items[i].DueDate < items[j].DueDate
	})
	if len(items) == 0 {
		return nil, nil
	}
	return &items[0], nil
}

func createAPITokenRecord(app core.App, userID string, name string) (*core.Record, string, error) {
	collection, err := app.FindCollectionByNameOrId("api_tokens")
	if err != nil {
		return nil, "", err
	}
	for attempt := 0; attempt < 3; attempt++ {
		plainToken, err := newPlainAPIToken()
		if err != nil {
			return nil, "", err
		}
		record := core.NewRecord(collection)
		record.Set("user", userID)
		record.Set("name", name)
		record.Set("tokenHash", hashPublicAPIToken(plainToken))
		record.Set("tokenPrefix", plainToken[:publicAPITokenPrefixLength])
		record.Set("scopes", []string{"read"})
		record.Set("lastUsedAt", "")
		if err := app.Save(record); err != nil {
			if attempt < 2 && strings.Contains(err.Error(), "idx_api_tokens_token_hash_unique") {
				continue
			}
			return nil, "", err
		}
		return record, plainToken, nil
	}
	return nil, "", errors.New("api token collision")
}

func findAPITokenForUser(app core.App, userID string, id string) (*core.Record, error) {
	record, err := app.FindFirstRecordByFilter(
		"api_tokens",
		"id = {:id} && user = {:user}",
		dbx.Params{"id": id, "user": userID},
	)
	if err != nil {
		return nil, err
	}
	return record, nil
}

func authenticatePublicAPIRequest(app core.App, e *core.RequestEvent) (publicAPIAuthContext, error) {
	token := bearerTokenFromRequest(e.Request)
	if !publicAPITokenRe.MatchString(token) {
		return publicAPIAuthContext{}, sql.ErrNoRows
	}
	hash := hashPublicAPIToken(token)
	record, err := app.FindFirstRecordByFilter(
		"api_tokens",
		"tokenHash = {:tokenHash}",
		dbx.Params{"tokenHash": hash},
	)
	if err != nil {
		return publicAPIAuthContext{}, sql.ErrNoRows
	}
	scopes, err := apiTokenScopesFromValue(record.Get("scopes"))
	if err != nil || !apiTokenHasReadScope(scopes) {
		return publicAPIAuthContext{}, sql.ErrNoRows
	}
	userID := record.GetString("user")
	user, err := app.FindRecordById("users", userID)
	if err != nil || user.GetBool("banned") {
		return publicAPIAuthContext{}, sql.ErrNoRows
	}
	// Public API token 是独立 bearer：成功鉴权后只更新 lastUsedAt，不创建或复用登录 session。
	record.Set("lastUsedAt", time.Now().UTC().Format(time.RFC3339Nano))
	if err := app.Save(record); err != nil {
		return publicAPIAuthContext{}, err
	}
	return publicAPIAuthContext{UserID: userID, Scopes: []string{"read"}}, nil
}

func publicAPIAuthError(e *core.RequestEvent, err error) error {
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return e.InternalServerError(serverText(requestLocale(e.Request), "common.internalError"), err)
	}
	setPublicAPIHeaders(e.Response.Header())
	return apiErrorJSON(e, http.StatusUnauthorized, "PUBLIC_API_UNAUTHORIZED", serverText(requestLocale(e.Request), "auth.loginRequired"), nil)
}

func findSubscriptionForPublicAPI(app core.App, userID string, id string) (*core.Record, error) {
	return app.FindFirstRecordByFilter(
		"subscriptions",
		"id = {:id} && user = {:user}",
		dbx.Params{"id": id, "user": userID},
	)
}

func publicAPIDueItemsFromRecords(rows []*core.Record, today string, through string) []publicAPIDueItem {
	items := []publicAPIDueItem{}
	for _, row := range rows {
		if dueType := publicAPIDueType(row, today, through); dueType != "" {
			items = append(items, publicAPIDueItem{
				DueDate:      publicAPIDueDate(row, dueType),
				DueType:      dueType,
				Subscription: subscriptionAPIFromRecord(row),
			})
		}
	}
	return items
}

func publicAPIDueType(row *core.Record, today string, through string) string {
	if row.GetString("status") == "trial" {
		if date := strings.TrimSpace(row.GetString("trialEndDate")); date >= today && date <= through {
			return "trial"
		}
	}
	date := strings.TrimSpace(row.GetString("nextBillingDate"))
	if date < today || date > through {
		return ""
	}
	if row.GetString("billingCycle") == "one-time" {
		if row.GetInt("oneTimeTermCount") <= 0 {
			return ""
		}
		return "expiry"
	}
	return "renewal"
}

func publicAPIDueDate(row *core.Record, dueType string) string {
	if dueType == "trial" {
		return row.GetString("trialEndDate")
	}
	return row.GetString("nextBillingDate")
}

func apiTokenDTOFromRecord(record *core.Record) apiTokenDTO {
	return apiTokenDTO{
		ID:          record.Id,
		Name:        record.GetString("name"),
		TokenPrefix: record.GetString("tokenPrefix"),
		Scopes:      []string{"read"},
		CreatedAt:   recordTimeString(record, "created"),
		LastUsedAt:  optionalRecordString(record, "lastUsedAt"),
	}
}

func optionalRecordString(record *core.Record, field string) *string {
	value := strings.TrimSpace(record.GetString(field))
	if value == "" {
		return nil
	}
	return &value
}

func apiTokenScopesFromValue(value interface{}) ([]string, error) {
	data, err := jsonBytesFromValue(value)
	if err != nil || len(strings.TrimSpace(string(data))) == 0 {
		return nil, err
	}
	var scopes []string
	if err := json.Unmarshal(data, &scopes); err != nil {
		return nil, err
	}
	return scopes, nil
}

func apiTokenHasReadScope(scopes []string) bool {
	for _, scope := range scopes {
		if scope == "read" {
			return true
		}
	}
	return false
}

func newPlainAPIToken() (string, error) {
	data := make([]byte, publicAPITokenRandomBytes)
	if _, err := rand.Read(data); err != nil {
		return "", err
	}
	return publicAPITokenPrefix + base64.RawURLEncoding.EncodeToString(data), nil
}

func hashPublicAPIToken(token string) string {
	sum := sha256.Sum256([]byte(token))
	return base64.RawURLEncoding.EncodeToString(sum[:])
}

func bearerTokenFromRequest(request *http.Request) string {
	header := strings.TrimSpace(request.Header.Get("Authorization"))
	match := bearerTokenHeaderRe.FindStringSubmatch(header)
	if len(match) != 2 {
		return ""
	}
	return strings.TrimSpace(match[1])
}

func setPublicAPIHeaders(headers http.Header) {
	headers.Set("Cache-Control", "no-store")
	headers.Set("X-Content-Type-Options", "nosniff")
}
